### PR TITLE
[pytree] add `tree_iter` function and fix opcode `YIELD_FROM` and `SEND`

### PR DESCRIPTION
**Pull request by XuehaiPan**
Reland #120155

Fix opcode `YIELD_FROM` and `SEND` opcode for dynamo.

Remove the guard return after `yield`.

```python
def func(...):
    if cond:
        yield something
        return

    # do something else
```

to

```python
def func(...):
    if cond:
        yield something
    else:
        # do something else
```

------

Original comment:

This PR adds a new function `tree_iter` that lazily iterates over the tree leaves. It is different than the `tree_leaves` function while the latter traversal the whole tree first to build a list of leaves.

```python
for leaf in tree_iter(tree):
    ...
```

is much more efficient than:

```python
for leaf in tree_leaves(tree):
    ...
```

where `tree_leaves(tree)` is `list(tree_iter(tree))`.

cc @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

**Comment by vmoens** [comment]
I think you can keep tree_iter but just rewriter it as in https://github.com/pytorch/pytorch/pull/120155#issuecomment-1955606472

**Comment by huydhn** [comment]
I'm importing the diff and will re-run the failed tests internally to confirm they are fixed.

**Comment by huydhn** [comment]
Just to confirm that the torchrec tests pass with the current PR

**Comment by XuehaiPan** [comment]
> I think you can keep tree_iter but just rewrite it as in [#120155 (comment)](https://github.com/pytorch/pytorch/pull/120155#issuecomment-1955606472)

I updated the PR and description. Let's see how it goes.

**Review by zou3519 — APPROVED** [review]
@huydhn I don't have context over why this was reverted. Could you check to see if this is landable?

File: test/dynamo/test_higher_order_ops.py [hunk 1]
```diff
@@ -4512,29 +4512,29 @@ def forward(self, L_x_ : torch.Tensor, L_y_ : torch.Tensor):
         dual = _unpack_dual_1[1];  _unpack_dual_1 = None
 
         child_4 = torch._C._functorch._unwrap_for_grad(primal, 2);  primal = None
-        child_5 = torch._C._functorch._unwrap_for_grad(primal_1, 2);  primal_1 = None
+        _unwrap_for_grad_1 = torch._C._functorch._unwrap_for_grad(primal_1, 2);  primal_1 = None
```
Comment (XuehaiPan) [inline test/dynamo/test_higher_order_ops.py:4515]: NOTE: This variable is not used in the generated code.

File: torch/_dynamo/symbolic_convert.py [hunk 2]
```diff
@@ -2649,14 +2649,14 @@ def YIELD_FROM(self, inst):
...
                 val = tos.next_variable(self)
+            except (StopIteration, exc.UserStopIteration):
+                self.pop()
+                return
+            else:
                 self.push(val)
                 # TODO(voz): Unclear if we need the push None in YIELD_VALUE?
                 self.YIELD_VALUE(inst)
                 self.pop()
-                self.push(tos)
```
Comment (XuehaiPan) [inline torch/_dynamo/symbolic_convert.py:2656]: `tos` is not popped from the stack. So we should not push it into the stack again.

File: torch/_dynamo/symbolic_convert.py [hunk 3]
```diff
@@ -2649,14 +2649,14 @@ def YIELD_FROM(self, inst):
                 return
             try:
                 val = tos.next_variable(self)
+            except (StopIteration, exc.UserStopIteration):
+                self.pop()
```
Comment (XuehaiPan) [inline torch/_dynamo/symbolic_convert.py:2653]: Pop-out `tos` when it is empty.

**Review by jansel — CHANGES_REQUESTED** [review]
failing tests?

**Comment by XuehaiPan** [comment]
I have fixed the implementation for opcode `YIELD_FROM` and `SEND`. Now all tests pass.

File: torch/_dynamo/symbolic_convert.py [hunk 4]
```diff
@@ -2641,33 +2641,46 @@ def GET_YIELD_FROM_ITER(self, inst):
...
+            # lifted the `unimplemented("generator")` in frame conversion. This codepath handles
+            # subgenerator and lines up with this line in Python 3.10
+            # https://github.com/python/cpython/blob/3.10/Python/ceval.c#L2599
+            unimplemented("Unreachable sub-generator code")
+
+        try:
+            val = tos.next_variable(self)
+        except (StopIteration, exc.UserStopIteration) as ex:
+            self.pop()
+            self.push(ConstantVariable.create(ex.value))
```
Comment (jansel) [inline torch/_dynamo/symbolic_convert.py:2661]: I think this will fail if the user code raises a StopIteration:
```
>>> torch._dynamo.exc.UserStopIteration().value
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'UserStopIteration' object has no attribute 'value'
```

File: torch/_dynamo/symbolic_convert.py [hunk 5]
```diff
@@ -2641,33 +2641,46 @@ def GET_YIELD_FROM_ITER(self, inst):
...
         if isinstance(tos, ListIteratorVariable):
             if isinstance(val, ConstantVariable) and val.value is None:
-                self.push(val)
-                self.instruction_pointer = self.indexof[inst.target]
+                try:
+                    val = tos.next_variable(self)
+                except (StopIteration, exc.UserStopIteration) as ex:
+                    if sys.version_info < (3, 12):
+                        self.pop()  # Python 3.12 uses new opcode END_SEND
+                    self.push(ConstantVariable.create(ex.value))
```
Comment (jansel) [inline torch/_dynamo/symbolic_convert.py:2680]: Same

File: torch/_dynamo/symbolic_convert.py [hunk 6]
```diff
@@ -2641,33 +2641,46 @@ def GET_YIELD_FROM_ITER(self, inst):
...
+                    self.instruction_pointer += inst.arg
```
Comment (jansel) [inline torch/_dynamo/symbolic_convert.py:2681]: In the past there have been some changes to how bytecodes are indexed between Python versions, so I think using `inst.target` is more robust.  What is the reason for this change?

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/symbolic_convert.py (hunk 4 above)
Comment (XuehaiPan) [reply torch/_dynamo/symbolic_convert.py:2661]: I add the `value` attribute to the class `UserStopIteration` in this PR.

Or we can ignore the value of the exception and push `ConstantVariable.create(None)`.

> if the user code raises a StopIteration

Can you give some hints on how can we test this? `StopIteration` will convert to `RuntimeError`.

```python
In [1]: def gen():
   ...:     yield 1
   ...:     yield 2
   ...:     raise StopIteration(3)
   ...:     

In [2]: list(gen())
StopIteration: 3

The above exception was the direct cause of the following exception:

RuntimeError: generator raised StopIteration
```

File: torch/_dynamo/symbolic_convert.py (hunk 6 above)
Comment (XuehaiPan) [reply torch/_dynamo/symbolic_convert.py:2681]: I changed the implementation based on [Python documentation opcode `SEND(delta)`](https://docs.python.org/3.12/library/dis.html#opcode-SEND):

> `SEND(delta)`
> Equivalent to `STACK[-1] = STACK[-2].send(STACK[-1])`. Used in yield from and await statements.
> 
> If the call raises [StopIteration](https://docs.python.org/3.12/library/exceptions.html#StopIteration), pop the top value from the stack, push the exception’s value attribute, and increment the bytecode counter by `delta`.
> 
> New in version 3.11.

And the CPython implementation for opcode `SEND`:

https://github.com/python/cpython/blob/3bc0d2b851bfa38bb3614d56bd9b83b1c651ea63/Python/ceval.c#L2617

```c
if (gen_status == PYGEN_RETURN) {
    assert(retval != NULL);
    Py_DECREF(receiver);
    SET_TOP(retval);
    JUMPBY(oparg);
    DISPATCH();
}
```

Comment (XuehaiPan) [reply torch/_dynamo/symbolic_convert.py]: I changed it back to use `self.instruction_pointer = self.indexof[inst.target]`.

**Comment by XuehaiPan** [comment]
Test failures are unrelated.

File: torch/_dynamo/symbolic_convert.py [hunk 7]
```diff
@@ -1121,8 +1121,18 @@ def load_builtin(self, inst):
...
-        self.instruction_pointer = self.indexof[inst.target]
+    def jump(self, inst, offset: Optional[int] = None):
+        # Jump to `inst.target` or jump by `offset`
+        if inst.target is not None:
+            assert offset is None
+            ip = self.indexof[inst.target]
+        else:
+            assert isinstance(self.instruction_pointer, int)
+            assert offset is not None
+            ip = self.instruction_pointer + offset
```
Comment (jansel) [inline torch/_dynamo/symbolic_convert.py:1132]: Add an `assert sys.version_info >= (3, 10)`

This is the change I am worried about here ([source](https://docs.python.org/3/library/dis.html))
```
Changed in version 3.10: The argument of jump, exception handling and loop instructions is now the instruction offset rather than the byte offset.
```

In prior versions of python you would need to divide the offset by 2.  So if we ever hit this code in Python 3.8/3.9 it will result in some very weird bugs.

More fundamentally, why isn't `inst.target` being set properly?  Can we just fix the code so that it is defined and this isn't needed?

File: torch/_dynamo/symbolic_convert.py (hunk 4 above)
Comment (jansel) [reply torch/_dynamo/symbolic_convert.py:2676]: You can test this with something like:
```
class MyIter:
  def __iter__(self):
     return self
  def __next__(self):
     raise StopIteration
```

When we trace through `__next__` we transform `StopIteration` to `UserStopIteration`.

**Review by jansel — CHANGES_REQUESTED** [review]
Please add some tests to this.

File: torch/_dynamo/symbolic_convert.py (hunk 7 above)
Comment (jansel) [reply torch/_dynamo/symbolic_convert.py:1132]: Also, we will need to assert on the opcode.  There are functions like JUMP_ABSOLUTE/JUMP_BACKWARD that would use a different formula to compute the new `ip`.

Comment (XuehaiPan) [reply torch/_dynamo/symbolic_convert.py]: I reverted this change.

File: test/dynamo/test_misc.py [hunk 8]
```diff
@@ -8760,6 +8760,34 @@ def yield_from_gen():
...
+                yield from MyIter(seq)
+
+            return [i for i in gen(seq)]
+
+        seq = [torch.randn([2, 3]) for _ in range(3)]
+        eager = yield_from_iter_fn(seq)
+        counter = CompileCounter()
+        compiled = torch._dynamo.optimize(counter)(yield_from_iter_fn)(seq)
+        self.assertEqual(eager, compiled)
+        self.assertEqual(counter.frame_count, 0)
```
Comment (XuehaiPan) [inline test/dynamo/test_misc.py:8789]: The frame count is 0 because a graph break occurred here:

https://github.com/pytorch/pytorch/blob/8aa08b8b9d1fab2a13dc5fbda74c553cb2a08729/torch/_dynamo/convert_frame.py#L349-L350

during frame conversion.

The graph capture process is successful. No attribute error is raised for `UserStopIteration.value`.

**Review by jansel — APPROVED** [review]

**Comment by XuehaiPan** [comment]
@huydhn Could you take a look at this?

**Comment by jansel** [comment]
Not sure what is going on here.   Somehow `Meta Internal-Only Changes Check` thinks this is linked to a phabricator diff.

One fix might be creating a new version of this PR.

**Comment by XuehaiPan** [comment]
I split this PR into:

- #123912
- #123913
